From 68e2c2a807b9799420fecf680086d6e66e582128 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sun, 9 Feb 2014 18:44:25 +0000
Subject: [PATCH 17/38] Add gettext and libiconv as a companion lib

.. it's needed for the RPC stuff in {e}glibc.

Building both is avoided on Linux and building
iconv is avoided on Darwin.
---
 config/companion_libs.in                           |  27 ++
 config/companion_libs/gettext.in                   |  19 ++
 config/companion_libs/libiconv.in                  |  19 ++
 config/libc/eglibc.in                              |   1 +
 config/libc/glibc.in                               |   1 +
 ...ix-linker-error-redefinition-of-vasprintf.patch |  31 ++
 ...ix-linker-error-redefinition-of-vasprintf.patch |  31 ++
 ...-Woe32-link-errors-when-compiling-with-O0.patch | 372 +++++++++++++++++++++
 ...rclib_stdio.in.h-remove-gets-declarations.patch |  30 ++
 scripts/build/companion_libs/320-libiconv.sh       | 107 ++++++
 scripts/build/companion_libs/330-gettext.sh        | 131 ++++++++
 scripts/build/libc/glibc-eglibc.sh-common          |  18 +-
 12 files changed, 783 insertions(+), 4 deletions(-)
 create mode 100644 config/companion_libs/gettext.in
 create mode 100644 config/companion_libs/libiconv.in
 create mode 100644 patches/gettext/0.18.2/110-Fix-linker-error-redefinition-of-vasprintf.patch
 create mode 100644 patches/gettext/0.18.3.1/110-Fix-linker-error-redefinition-of-vasprintf.patch
 create mode 100644 patches/gettext/0.18.3.1/120-Fix-Woe32-link-errors-when-compiling-with-O0.patch
 create mode 100644 patches/libiconv/1.14/100-srclib_stdio.in.h-remove-gets-declarations.patch
 create mode 100644 scripts/build/companion_libs/320-libiconv.sh
 create mode 100644 scripts/build/companion_libs/330-gettext.sh

diff --git a/config/companion_libs.in b/config/companion_libs.in
index db1c600..83d9f14 100644
--- a/config/companion_libs.in
+++ b/config/companion_libs.in
@@ -57,6 +57,17 @@ config OPENSSL_NEEDED
     select OPENSSL
     select COMPLIBS_NEEDED
 
+config LIBICONV_NEEDED
+    bool
+    select LIBICONV
+    select COMPLIBS_NEEDED
+
+config GETTEXT_NEEDED
+    bool
+    select GETTEXT
+    select LIBICONV_NEEDED
+    select COMPLIBS_NEEDED
+
 config LIBUUID_NEEDED
     bool
     select LIBUUID
@@ -110,6 +121,14 @@ config OPENSSL
     bool
     select COMPLIBS
 
+config LIBICONV
+    bool
+    select COMPLIBS
+
+config GETTEXT
+    bool
+    select COMPLIBS
+
 config LIBUUID
     bool
     select COMPLIBS
@@ -146,6 +165,14 @@ if OPENSSL
 source "config/companion_libs/openssl.in"
 endif
 
+if LIBICONV
+source "config/companion_libs/libiconv.in"
+endif
+
+if GETTEXT
+source "config/companion_libs/gettext.in"
+endif
+
 if LLVM
 source "config/companion_libs/llvm.in"
 endif
diff --git a/config/companion_libs/gettext.in b/config/companion_libs/gettext.in
new file mode 100644
index 0000000..e600b66
--- /dev/null
+++ b/config/companion_libs/gettext.in
@@ -0,0 +1,19 @@
+# gettext options
+
+choice
+    bool
+    prompt "gettext version"
+# Don't remove next line
+# CT_INSERT_VERSION_BELOW
+
+config GETTEXT_V_0_18_3_1
+    bool
+    prompt "0.18.3.1"
+
+endchoice
+
+config GETTEXT_VERSION
+    string
+# Don't remove next line
+# CT_INSERT_VERSION_STRING_BELOW
+    default "0.18.3.1" if GETTEXT_V_0_18_3_1
diff --git a/config/companion_libs/libiconv.in b/config/companion_libs/libiconv.in
new file mode 100644
index 0000000..361d34e
--- /dev/null
+++ b/config/companion_libs/libiconv.in
@@ -0,0 +1,19 @@
+# libiconv options
+
+choice
+    bool
+    prompt "libiconv version"
+# Don't remove next line
+# CT_INSERT_VERSION_BELOW
+
+config LIBICONV_V_1_14
+    bool
+    prompt "1.14"
+
+endchoice
+
+config LIBICONV_VERSION
+    string
+# Don't remove next line
+# CT_INSERT_VERSION_STRING_BELOW
+    default "1.14" if LIBICONV_V_1_14
diff --git a/config/libc/eglibc.in b/config/libc/eglibc.in
index 8dc4e44..cc2e8d1 100644
--- a/config/libc/eglibc.in
+++ b/config/libc/eglibc.in
@@ -4,6 +4,7 @@
 ##
 ## select LIBC_SUPPORT_NPTL
 ## select CC_CORE_PASSES_NEEDED
+## select GETTEXT_NEEDED
 ##
 ## help EGLIBC (Embedded GLIBC) is a variant of the standard GNU GLIBC
 ## help that is designed to work well on embedded systems.  EGLIBC strives
diff --git a/config/libc/glibc.in b/config/libc/glibc.in
index 6cc3852..152513f 100644
--- a/config/libc/glibc.in
+++ b/config/libc/glibc.in
@@ -4,6 +4,7 @@
 ##
 ## select LIBC_SUPPORT_NPTL
 ## select CC_CORE_PASSES_NEEDED
+## select GETTEXT_NEEDED
 ##
 ## help The de-facto standard for Linux distributions.
 ## help Feature-rich, but large...  Most usefull for desktop-like systems.
diff --git a/patches/gettext/0.18.2/110-Fix-linker-error-redefinition-of-vasprintf.patch b/patches/gettext/0.18.2/110-Fix-linker-error-redefinition-of-vasprintf.patch
new file mode 100644
index 0000000..a5287b3
--- /dev/null
+++ b/patches/gettext/0.18.2/110-Fix-linker-error-redefinition-of-vasprintf.patch
@@ -0,0 +1,31 @@
+From a76649dae62768d0af7017b3fc0ca5f891588c78 Mon Sep 17 00:00:00 2001
+From: Andoni Morales Alastruey <ylatuya@gmail.com>
+Date: Wed, 29 Feb 2012 10:44:43 +0100
+Subject: [PATCH] Fix linker error: redefinition of vasprintf
+
+This might not be the best patch, but it works for us
+The link error was:
+.libs/autosprintf.o:autosprintf.cc:(.text$vasprintf[_vasprintf]+0x0): multiple definition of `_vasprintf'
+.libs/lib-asprintf.o:lib-asprintf.c:(.text+0x4621): first defined here
+---
+ gettext-runtime/libasprintf/autosprintf.cc | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/gettext-runtime/libasprintf/autosprintf.cc b/gettext-runtime/libasprintf/autosprintf.cc
+index ca318f7..2526210 100644
+--- a/gettext-runtime/libasprintf/autosprintf.cc
++++ b/gettext-runtime/libasprintf/autosprintf.cc
+@@ -21,8 +21,10 @@
+    This must come before <config.h> because <config.h> may include
+    <features.h>, and once <features.h> has been included, it's too late.  */
+ #ifndef _GNU_SOURCE
++#ifndef _WIN32
+ # define _GNU_SOURCE    1
+ #endif
++#endif
+ 
+ /* Specification.  */
+ #include "autosprintf.h"
+-- 
+1.8.4
+
diff --git a/patches/gettext/0.18.3.1/110-Fix-linker-error-redefinition-of-vasprintf.patch b/patches/gettext/0.18.3.1/110-Fix-linker-error-redefinition-of-vasprintf.patch
new file mode 100644
index 0000000..dacdfb1
--- /dev/null
+++ b/patches/gettext/0.18.3.1/110-Fix-linker-error-redefinition-of-vasprintf.patch
@@ -0,0 +1,31 @@
+From a76649dae62768d0af7017b3fc0ca5f891588c78 Mon Sep 17 00:00:00 2001
+From: Andoni Morales Alastruey <ylatuya@gmail.com>
+Date: Wed, 29 Feb 2012 10:44:43 +0100
+Subject: [PATCH] Fix linker error: redefinition of vasprintf
+
+This might not be the best patch, but it works for us
+The link error was:
+.libs/autosprintf.o:autosprintf.cc:(.text$vasprintf[_vasprintf]+0x0): multiple definition of `_vasprintf'
+.libs/lib-asprintf.o:lib-asprintf.c:(.text+0x4621): first defined here
+---
+ gettext-runtime/libasprintf/autosprintf.cc | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/gettext-runtime/libasprintf/autosprintf.cc b/gettext-runtime/libasprintf/autosprintf.cc
+index ca318f7..2526210 100644
+--- a/gettext-runtime/libasprintf/autosprintf.cc
++++ b/gettext-runtime/libasprintf/autosprintf.cc
+@@ -19,8 +19,10 @@
+    This must come before <config.h> because <config.h> may include
+    <features.h>, and once <features.h> has been included, it's too late.  */
+ #ifndef _GNU_SOURCE
++#ifndef _WIN32
+ # define _GNU_SOURCE    1
+ #endif
++#endif
+ 
+ /* Specification.  */
+ #include "autosprintf.h"
+-- 
+1.8.4
+
diff --git a/patches/gettext/0.18.3.1/120-Fix-Woe32-link-errors-when-compiling-with-O0.patch b/patches/gettext/0.18.3.1/120-Fix-Woe32-link-errors-when-compiling-with-O0.patch
new file mode 100644
index 0000000..24e75aa
--- /dev/null
+++ b/patches/gettext/0.18.3.1/120-Fix-Woe32-link-errors-when-compiling-with-O0.patch
@@ -0,0 +1,372 @@
+From d4ecf6f15ad7a428786df2efdc88b03be0a4fdbb Mon Sep 17 00:00:00 2001
+From: Daiki Ueno <ueno@gnu.org>
+Date: Thu, 17 Jan 2013 18:33:40 +0900
+Subject: [PATCH] Fix Woe32 link errors when compiling with -O0.
+
+---
+ gettext-tools/src/Makefile.am                  | 24 ++++++++++++++++++++++--
+ gettext-tools/src/color.c                      |  1 +
+ gettext-tools/woe32dll/c++color.cc             |  1 +
+ gettext-tools/woe32dll/c++file-ostream.cc      |  2 ++
+ gettext-tools/woe32dll/c++html-ostream.cc      |  1 +
+ gettext-tools/woe32dll/c++styled-ostream.cc    |  1 +
+ gettext-tools/woe32dll/c++term-ostream.cc      |  1 +
+ gettext-tools/woe32dll/c++write-catalog.cc     |  1 +
+ gettext-tools/woe32dll/c++write-po.cc          |  1 +
+ gettext-tools/woe32dll/c++write-properties.cc  |  1 +
+ gettext-tools/woe32dll/c++write-stringtable.cc |  1 +
+ gnulib-local/modules/file-ostream              |  4 ++++
+ gnulib-local/modules/html-ostream              |  4 ++++
+ gnulib-local/modules/ostream                   |  4 ++++
+ gnulib-local/modules/styled-ostream            |  4 ++++
+ gnulib-local/modules/term-ostream              |  4 ++++
+ 16 files changed, 53 insertions(+), 2 deletions(-)
+ create mode 100644 gettext-tools/woe32dll/c++color.cc
+ create mode 100644 gettext-tools/woe32dll/c++file-ostream.cc
+ create mode 100644 gettext-tools/woe32dll/c++html-ostream.cc
+ create mode 100644 gettext-tools/woe32dll/c++styled-ostream.cc
+ create mode 100644 gettext-tools/woe32dll/c++term-ostream.cc
+ create mode 100644 gettext-tools/woe32dll/c++write-catalog.cc
+ create mode 100644 gettext-tools/woe32dll/c++write-po.cc
+ create mode 100644 gettext-tools/woe32dll/c++write-properties.cc
+ create mode 100644 gettext-tools/woe32dll/c++write-stringtable.cc
+
+Additional fix (COLOR_SOURCE) by Ray Donnelly <mingw.android@gmail.com>
+
+diff -urN a/gettext-tools/src/Makefile.am b/gettext-tools/src/Makefile.am
+--- a/gettext-tools/src/Makefile.am	2013-08-15 17:15:56.000000000 +0100
++++ b/gettext-tools/src/Makefile.am	2013-12-12 23:49:22.198570891 +0000
+@@ -141,15 +141,36 @@
+   format-lua.c \
+   format-javascript.c
+ 
++if !WOE32DLL
++COLOR_SOURCE = color.c
++else
++COLOR_SOURCE = ../woe32dll/c++color.cc
++endif
++
++if !WOE32DLL
++OUTPUT_SOURCE = \
++  write-catalog.c \
++  write-po.c \
++  write-properties.c \
++  write-stringtable.c
++else
++OUTPUT_SOURCE = \
++  ../woe32dll/c++write-catalog.cc \
++  ../woe32dll/c++write-po.cc \
++  ../woe32dll/c++write-properties.cc \
++  ../woe32dll/c++write-stringtable.cc
++endif
++
+ # libgettextsrc contains all code that is needed by at least two programs.
+ libgettextsrc_la_SOURCES = \
+ $(COMMON_SOURCE) read-catalog.c \
+-color.c write-catalog.c write-properties.c write-stringtable.c write-po.c \
++$(COLOR_SOURCE) $(OUTPUT_SOURCE) \
+ msgl-ascii.c msgl-iconv.c msgl-equal.c msgl-cat.c msgl-header.c msgl-english.c \
+ msgl-check.c file-list.c msgl-charset.c po-time.c plural-exp.c plural-eval.c \
+ plural-table.c \
+ $(FORMAT_SOURCE)
+ 
++
+ # msggrep needs pattern matching.
+ LIBGREP = ../libgrep/libgrep.a
+ 
+diff -urN a/gettext-tools/src/Makefile.in b/gettext-tools/src/Makefile.in
+--- a/gettext-tools/src/Makefile.in	2013-08-15 17:31:16.000000000 +0100
++++ b/gettext-tools/src/Makefile.in	2013-12-12 23:53:30.512410954 +0000
+@@ -377,16 +377,20 @@
+ am__libgettextsrc_la_SOURCES_DIST = message.c po-error.c po-xerror.c \
+ 	read-catalog-abstract.c po-lex.c po-gram-gen.y po-charset.c \
+ 	read-po.c read-properties.c read-stringtable.c open-catalog.c \
+-	dir-list.c str-list.c read-catalog.c color.c write-catalog.c \
+-	write-properties.c write-stringtable.c write-po.c msgl-ascii.c \
+-	msgl-iconv.c msgl-equal.c msgl-cat.c msgl-header.c \
+-	msgl-english.c msgl-check.c file-list.c msgl-charset.c \
+-	po-time.c plural-exp.c plural-eval.c plural-table.c format.c \
+-	format-invalid.h format-c.c format-c-parse.h format-sh.c \
+-	format-python.c format-python-brace.c format-lisp.c \
+-	format-elisp.c format-librep.c format-scheme.c format-java.c \
+-	format-csharp.c format-awk.c format-pascal.c format-ycp.c \
+-	format-tcl.c format-perl.c format-perl-brace.c format-php.c \
++	dir-list.c str-list.c read-catalog.c color.c \
++	../woe32dll/c++color.cc write-catalog.c write-po.c \
++	write-properties.c write-stringtable.c \
++	../woe32dll/c++write-catalog.cc ../woe32dll/c++write-po.cc \
++	../woe32dll/c++write-properties.cc \
++	../woe32dll/c++write-stringtable.cc msgl-ascii.c msgl-iconv.c \
++	msgl-equal.c msgl-cat.c msgl-header.c msgl-english.c \
++	msgl-check.c file-list.c msgl-charset.c po-time.c plural-exp.c \
++	plural-eval.c plural-table.c format.c format-invalid.h \
++	format-c.c format-c-parse.h format-sh.c format-python.c \
++	format-python-brace.c format-lisp.c format-elisp.c \
++	format-librep.c format-scheme.c format-java.c format-csharp.c \
++	format-awk.c format-pascal.c format-ycp.c format-tcl.c \
++	format-perl.c format-perl-brace.c format-php.c \
+ 	format-gcc-internal.c format-gfc-internal.c format-qt.c \
+ 	format-qt-plural.c format-kde.c format-boost.c format-lua.c \
+ 	format-javascript.c ../woe32dll/c++format.cc \
+@@ -400,7 +404,17 @@
+ 	libgettextsrc_la-read-stringtable.lo \
+ 	libgettextsrc_la-open-catalog.lo libgettextsrc_la-dir-list.lo \
+ 	libgettextsrc_la-str-list.lo
+-@WOE32DLL_FALSE@am__objects_2 = libgettextsrc_la-format.lo \
++@WOE32DLL_FALSE@am__objects_2 = libgettextsrc_la-color.lo
++@WOE32DLL_TRUE@am__objects_2 = libgettextsrc_la-c++color.lo
++@WOE32DLL_FALSE@am__objects_3 = libgettextsrc_la-write-catalog.lo \
++@WOE32DLL_FALSE@	libgettextsrc_la-write-po.lo \
++@WOE32DLL_FALSE@	libgettextsrc_la-write-properties.lo \
++@WOE32DLL_FALSE@	libgettextsrc_la-write-stringtable.lo
++@WOE32DLL_TRUE@am__objects_3 = libgettextsrc_la-c++write-catalog.lo \
++@WOE32DLL_TRUE@	libgettextsrc_la-c++write-po.lo \
++@WOE32DLL_TRUE@	libgettextsrc_la-c++write-properties.lo \
++@WOE32DLL_TRUE@	libgettextsrc_la-c++write-stringtable.lo
++@WOE32DLL_FALSE@am__objects_4 = libgettextsrc_la-format.lo \
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-c.lo \
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-sh.lo \
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-python.lo \
+@@ -426,7 +440,7 @@
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-boost.lo \
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-lua.lo \
+ @WOE32DLL_FALSE@	libgettextsrc_la-format-javascript.lo
+-@WOE32DLL_TRUE@am__objects_2 = libgettextsrc_la-c++format.lo \
++@WOE32DLL_TRUE@am__objects_4 = libgettextsrc_la-c++format.lo \
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-c.lo \
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-sh.lo \
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-python.lo \
+@@ -452,21 +466,18 @@
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-boost.lo \
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-lua.lo \
+ @WOE32DLL_TRUE@	libgettextsrc_la-format-javascript.lo
+-@WOE32DLL_TRUE@am__objects_3 = libgettextsrc_la-gettextsrc-exports.lo
++@WOE32DLL_TRUE@am__objects_5 = libgettextsrc_la-gettextsrc-exports.lo
+ am_libgettextsrc_la_OBJECTS = $(am__objects_1) \
+-	libgettextsrc_la-read-catalog.lo libgettextsrc_la-color.lo \
+-	libgettextsrc_la-write-catalog.lo \
+-	libgettextsrc_la-write-properties.lo \
+-	libgettextsrc_la-write-stringtable.lo \
+-	libgettextsrc_la-write-po.lo libgettextsrc_la-msgl-ascii.lo \
++	libgettextsrc_la-read-catalog.lo $(am__objects_2) \
++	$(am__objects_3) libgettextsrc_la-msgl-ascii.lo \
+ 	libgettextsrc_la-msgl-iconv.lo libgettextsrc_la-msgl-equal.lo \
+ 	libgettextsrc_la-msgl-cat.lo libgettextsrc_la-msgl-header.lo \
+ 	libgettextsrc_la-msgl-english.lo \
+ 	libgettextsrc_la-msgl-check.lo libgettextsrc_la-file-list.lo \
+ 	libgettextsrc_la-msgl-charset.lo libgettextsrc_la-po-time.lo \
+ 	libgettextsrc_la-plural-exp.lo libgettextsrc_la-plural-eval.lo \
+-	libgettextsrc_la-plural-table.lo $(am__objects_2) \
+-	$(am__objects_3)
++	libgettextsrc_la-plural-table.lo $(am__objects_4) \
++	$(am__objects_5)
+ libgettextsrc_la_OBJECTS = $(am_libgettextsrc_la_OBJECTS)
+ PROGRAMS = $(bin_PROGRAMS) $(noinst_PROGRAMS)
+ am_hostname_OBJECTS = hostname-hostname.$(OBJEXT)
+@@ -1902,14 +1913,27 @@
+ @WOE32DLL_TRUE@	format-gcc-internal.c format-gfc-internal.c \
+ @WOE32DLL_TRUE@	format-qt.c format-qt-plural.c format-kde.c \
+ @WOE32DLL_TRUE@	format-boost.c format-lua.c format-javascript.c
++@WOE32DLL_FALSE@COLOR_SOURCE = color.c
++@WOE32DLL_TRUE@COLOR_SOURCE = ../woe32dll/c++color.cc
++@WOE32DLL_FALSE@OUTPUT_SOURCE = \
++@WOE32DLL_FALSE@  write-catalog.c \
++@WOE32DLL_FALSE@  write-po.c \
++@WOE32DLL_FALSE@  write-properties.c \
++@WOE32DLL_FALSE@  write-stringtable.c
++
++@WOE32DLL_TRUE@OUTPUT_SOURCE = \
++@WOE32DLL_TRUE@  ../woe32dll/c++write-catalog.cc \
++@WOE32DLL_TRUE@  ../woe32dll/c++write-po.cc \
++@WOE32DLL_TRUE@  ../woe32dll/c++write-properties.cc \
++@WOE32DLL_TRUE@  ../woe32dll/c++write-stringtable.cc
++
+ 
+ # libgettextsrc contains all code that is needed by at least two programs.
+-libgettextsrc_la_SOURCES = $(COMMON_SOURCE) read-catalog.c color.c \
+-	write-catalog.c write-properties.c write-stringtable.c \
+-	write-po.c msgl-ascii.c msgl-iconv.c msgl-equal.c msgl-cat.c \
+-	msgl-header.c msgl-english.c msgl-check.c file-list.c \
+-	msgl-charset.c po-time.c plural-exp.c plural-eval.c \
+-	plural-table.c $(FORMAT_SOURCE) $(am__append_1)
++libgettextsrc_la_SOURCES = $(COMMON_SOURCE) read-catalog.c \
++	$(COLOR_SOURCE) $(OUTPUT_SOURCE) msgl-ascii.c msgl-iconv.c \
++	msgl-equal.c msgl-cat.c msgl-header.c msgl-english.c \
++	msgl-check.c file-list.c msgl-charset.c po-time.c plural-exp.c \
++	plural-eval.c plural-table.c $(FORMAT_SOURCE) $(am__append_1)
+ 
+ # msggrep needs pattern matching.
+ LIBGREP = ../libgrep/libgrep.a
+@@ -2476,15 +2500,15 @@
+ libgettextsrc_la-write-catalog.lo: write-catalog.c
+ 	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-write-catalog.lo `test -f 'write-catalog.c' || echo '$(srcdir)/'`write-catalog.c
+ 
++libgettextsrc_la-write-po.lo: write-po.c
++	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-write-po.lo `test -f 'write-po.c' || echo '$(srcdir)/'`write-po.c
++
+ libgettextsrc_la-write-properties.lo: write-properties.c
+ 	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-write-properties.lo `test -f 'write-properties.c' || echo '$(srcdir)/'`write-properties.c
+ 
+ libgettextsrc_la-write-stringtable.lo: write-stringtable.c
+ 	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-write-stringtable.lo `test -f 'write-stringtable.c' || echo '$(srcdir)/'`write-stringtable.c
+ 
+-libgettextsrc_la-write-po.lo: write-po.c
+-	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-write-po.lo `test -f 'write-po.c' || echo '$(srcdir)/'`write-po.c
+-
+ libgettextsrc_la-msgl-ascii.lo: msgl-ascii.c
+ 	$(AM_V_CC)$(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o libgettextsrc_la-msgl-ascii.lo `test -f 'msgl-ascii.c' || echo '$(srcdir)/'`msgl-ascii.c
+ 
+@@ -2980,6 +3004,21 @@
+ .cc.lo:
+ 	$(AM_V_CXX)$(LTCXXCOMPILE) -c -o $@ $<
+ 
++libgettextsrc_la-c++color.lo: ../woe32dll/c++color.cc
++	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++color.lo `test -f '../woe32dll/c++color.cc' || echo '$(srcdir)/'`../woe32dll/c++color.cc
++
++libgettextsrc_la-c++write-catalog.lo: ../woe32dll/c++write-catalog.cc
++	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++write-catalog.lo `test -f '../woe32dll/c++write-catalog.cc' || echo '$(srcdir)/'`../woe32dll/c++write-catalog.cc
++
++libgettextsrc_la-c++write-po.lo: ../woe32dll/c++write-po.cc
++	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++write-po.lo `test -f '../woe32dll/c++write-po.cc' || echo '$(srcdir)/'`../woe32dll/c++write-po.cc
++
++libgettextsrc_la-c++write-properties.lo: ../woe32dll/c++write-properties.cc
++	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++write-properties.lo `test -f '../woe32dll/c++write-properties.cc' || echo '$(srcdir)/'`../woe32dll/c++write-properties.cc
++
++libgettextsrc_la-c++write-stringtable.lo: ../woe32dll/c++write-stringtable.cc
++	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++write-stringtable.lo `test -f '../woe32dll/c++write-stringtable.cc' || echo '$(srcdir)/'`../woe32dll/c++write-stringtable.cc
++
+ libgettextsrc_la-c++format.lo: ../woe32dll/c++format.cc
+ 	$(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(libgettextsrc_la_CPPFLAGS) $(CPPFLAGS) $(AM_CXXFLAGS) $(CXXFLAGS) -c -o libgettextsrc_la-c++format.lo `test -f '../woe32dll/c++format.cc' || echo '$(srcdir)/'`../woe32dll/c++format.cc
+ 
+diff -urN a/gettext-tools/src/color.c b/gettext-tools/src/color.c
+--- a/gettext-tools/src/color.c	2012-12-10 01:59:06.000000000 +0000
++++ b/gettext-tools/src/color.c	2013-12-12 23:49:00.401567136 +0000
+@@ -28,6 +28,7 @@
+ #include <sys/types.h>
+ #include <sys/stat.h>
+ 
++#include "ostream.h"
+ #include "term-ostream.h"
+ #include "xalloc.h"
+ #include "relocatable.h"
+diff -urN a/gettext-tools/woe32dll/c++color.cc b/gettext-tools/woe32dll/c++color.cc
+--- a/gettext-tools/woe32dll/c++color.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++color.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../src/color.c"
+diff -urN a/gettext-tools/woe32dll/c++file-ostream.cc b/gettext-tools/woe32dll/c++file-ostream.cc
+--- a/gettext-tools/woe32dll/c++file-ostream.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++file-ostream.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1,2 @@
++#include "../gnulib-lib/file-ostream.c"
++
+diff -urN a/gettext-tools/woe32dll/c++html-ostream.cc b/gettext-tools/woe32dll/c++html-ostream.cc
+--- a/gettext-tools/woe32dll/c++html-ostream.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++html-ostream.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../gnulib-lib/html-ostream.c"
+diff -urN a/gettext-tools/woe32dll/c++styled-ostream.cc b/gettext-tools/woe32dll/c++styled-ostream.cc
+--- a/gettext-tools/woe32dll/c++styled-ostream.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++styled-ostream.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../gnulib-lib/styled-ostream.c"
+diff -urN a/gettext-tools/woe32dll/c++term-ostream.cc b/gettext-tools/woe32dll/c++term-ostream.cc
+--- a/gettext-tools/woe32dll/c++term-ostream.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++term-ostream.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../gnulib-lib/term-ostream.c"
+diff -urN a/gettext-tools/woe32dll/c++write-catalog.cc b/gettext-tools/woe32dll/c++write-catalog.cc
+--- a/gettext-tools/woe32dll/c++write-catalog.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++write-catalog.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../src/write-catalog.c"
+diff -urN a/gettext-tools/woe32dll/c++write-po.cc b/gettext-tools/woe32dll/c++write-po.cc
+--- a/gettext-tools/woe32dll/c++write-po.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++write-po.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../src/write-po.c"
+diff -urN a/gettext-tools/woe32dll/c++write-properties.cc b/gettext-tools/woe32dll/c++write-properties.cc
+--- a/gettext-tools/woe32dll/c++write-properties.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++write-properties.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../src/write-properties.c"
+diff -urN a/gettext-tools/woe32dll/c++write-stringtable.cc b/gettext-tools/woe32dll/c++write-stringtable.cc
+--- a/gettext-tools/woe32dll/c++write-stringtable.cc	1970-01-01 01:00:00.000000000 +0100
++++ b/gettext-tools/woe32dll/c++write-stringtable.cc	2013-12-12 23:49:00.401567136 +0000
+@@ -0,0 +1 @@
++#include "../src/write-stringtable.c"
+diff -urN a/gnulib-local/modules/file-ostream b/gnulib-local/modules/file-ostream
+--- a/gnulib-local/modules/file-ostream	2012-12-10 01:59:06.000000000 +0000
++++ b/gnulib-local/modules/file-ostream	2013-12-12 23:49:00.401567136 +0000
+@@ -12,7 +12,11 @@
+ configure.ac:
+ 
+ Makefile.am:
++if !WOE32DLL
+ lib_SOURCES += file-ostream.c
++else
++lib_SOURCES += ../woe32dll/c++file-ostream.cc
++endif
+ # This is a Makefile rule that generates multiple files at once; see the
+ # automake documentation, node "Multiple Outputs", for details.
+ file-ostream.h : $(top_srcdir)/build-aux/moopp file-ostream.oo.h file-ostream.oo.c ostream.oo.h
+diff -urN a/gnulib-local/modules/html-ostream b/gnulib-local/modules/html-ostream
+--- a/gnulib-local/modules/html-ostream	2012-12-10 01:59:06.000000000 +0000
++++ b/gnulib-local/modules/html-ostream	2013-12-12 23:49:00.401567136 +0000
+@@ -15,7 +15,11 @@
+ configure.ac:
+ 
+ Makefile.am:
++if !WOE32DLL
+ lib_SOURCES += html-ostream.c
++else
++lib_SOURCES += ../woe32dll/c++html-ostream.cc
++endif
+ # This is a Makefile rule that generates multiple files at once; see the
+ # automake documentation, node "Multiple Outputs", for details.
+ html-ostream.h : $(top_srcdir)/build-aux/moopp html-ostream.oo.h html-ostream.oo.c ostream.oo.h
+diff -urN a/gnulib-local/modules/ostream b/gnulib-local/modules/ostream
+--- a/gnulib-local/modules/ostream	2012-12-10 01:59:06.000000000 +0000
++++ b/gnulib-local/modules/ostream	2013-12-12 23:49:00.401567136 +0000
+@@ -11,7 +11,11 @@
+ configure.ac:
+ 
+ Makefile.am:
++if !WOE32DLL
+ lib_SOURCES += ostream.c
++else
++lib_SOURCES += ../woe32dll/c++ostream.cc
++endif
+ # This is a Makefile rule that generates multiple files at once; see the
+ # automake documentation, node "Multiple Outputs", for details.
+ ostream.h : $(top_srcdir)/build-aux/moopp ostream.oo.h ostream.oo.c
+diff -urN a/gnulib-local/modules/styled-ostream b/gnulib-local/modules/styled-ostream
+--- a/gnulib-local/modules/styled-ostream	2012-12-10 01:59:06.000000000 +0000
++++ b/gnulib-local/modules/styled-ostream	2013-12-12 23:49:00.401567136 +0000
+@@ -11,7 +11,11 @@
+ configure.ac:
+ 
+ Makefile.am:
++if !WOE32DLL
+ lib_SOURCES += styled-ostream.c
++else
++lib_SOURCES += ../woe32dll/c++styled-ostream.cc
++endif
+ # This is a Makefile rule that generates multiple files at once; see the
+ # automake documentation, node "Multiple Outputs", for details.
+ styled-ostream.h : $(top_srcdir)/build-aux/moopp styled-ostream.oo.h styled-ostream.oo.c ostream.oo.h
+diff -urN a/gnulib-local/modules/term-ostream b/gnulib-local/modules/term-ostream
+--- a/gnulib-local/modules/term-ostream	2012-12-10 01:59:06.000000000 +0000
++++ b/gnulib-local/modules/term-ostream	2013-12-12 23:49:00.401567136 +0000
+@@ -22,7 +22,11 @@
+ gl_TERM_OSTREAM
+ 
+ Makefile.am:
++if !WOE32DLL
+ lib_SOURCES += term-ostream.c
++else
++lib_SOURCES += ../woe32dll/c++term-ostream.cc
++endif
+ # This is a Makefile rule that generates multiple files at once; see the
+ # automake documentation, node "Multiple Outputs", for details.
+ term-ostream.h : $(top_srcdir)/build-aux/moopp term-ostream.oo.h term-ostream.oo.c ostream.oo.h
diff --git a/patches/libiconv/1.14/100-srclib_stdio.in.h-remove-gets-declarations.patch b/patches/libiconv/1.14/100-srclib_stdio.in.h-remove-gets-declarations.patch
new file mode 100644
index 0000000..382c841
--- /dev/null
+++ b/patches/libiconv/1.14/100-srclib_stdio.in.h-remove-gets-declarations.patch
@@ -0,0 +1,30 @@
+diff -r -u libiconv-1.14/srclib/stdio.in.h.orig libiconv-1.14/srclib/stdio.in.h
+--- libiconv-1.14/srclib/stdio.in.h.orig	2013-02-22 13:52:46.336327969 -0600
++++ libiconv-1.14/srclib/stdio.in.h	2013-02-22 13:54:27.948207059 -0600
+@@ -679,22 +679,11 @@
+ # endif
+ #endif
+ 
+-#if @GNULIB_GETS@
+-# if @REPLACE_STDIO_READ_FUNCS@ && @GNULIB_STDIO_H_NONBLOCKING@
+-#  if !(defined __cplusplus && defined GNULIB_NAMESPACE)
+-#   undef gets
+-#   define gets rpl_gets
+-#  endif
+-_GL_FUNCDECL_RPL (gets, char *, (char *s) _GL_ARG_NONNULL ((1)));
+-_GL_CXXALIAS_RPL (gets, char *, (char *s));
+-# else
+-_GL_CXXALIAS_SYS (gets, char *, (char *s));
+-#  undef gets
+-# endif
+-_GL_CXXALIASWARN (gets);
+ /* It is very rare that the developer ever has full control of stdin,
+-   so any use of gets warrants an unconditional warning.  Assume it is
+-   always declared, since it is required by C89.  */
++   so any use of gets warrants an unconditional warning; besides, C11
++   removed it.  */
++#undef gets
++#if HAVE_RAW_DECL_GETS
+ _GL_WARN_ON_USE (gets, "gets is a security hole - use fgets instead");
+ #endif
+ 
diff --git a/scripts/build/companion_libs/320-libiconv.sh b/scripts/build/companion_libs/320-libiconv.sh
new file mode 100644
index 0000000..5d8443a
--- /dev/null
+++ b/scripts/build/companion_libs/320-libiconv.sh
@@ -0,0 +1,107 @@
+# Build script for libiconv
+
+do_libiconv_get() { :; }
+do_libiconv_extract() { :; }
+do_libiconv_for_build() { :; }
+do_libiconv_for_host() { :; }
+
+if [ "${CT_LIBICONV}" = "y" ]; then
+
+do_libiconv_get() {
+    CT_GetFile "libiconv-${CT_LIBICONV_VERSION}" \
+               http://ftp.gnu.org/pub/gnu/libiconv/
+}
+
+do_libiconv_extract() {
+    CT_Extract "libiconv-${CT_LIBICONV_VERSION}"
+    CT_Patch "libiconv" "${CT_LIBICONV_VERSION}"
+}
+
+# Build libiconv for running on build
+do_libiconv_for_build() {
+    local -a libiconv_opts
+
+    case "$CT_BUILD" in
+        *darwin*|*linux*)
+            return 0
+            ;;
+    esac
+
+    CT_DoStep INFO "Installing libiconv for build"
+    CT_mkdir_pushd "${CT_BUILD_DIR}/build-libiconv-build-${CT_BUILD}"
+
+    libiconv_opts+=( "host=${CT_BUILD}" )
+    libiconv_opts+=( "prefix=${CT_BUILDTOOLS_PREFIX_DIR}" )
+    libiconv_opts+=( "cflags=${CT_CFLAGS_FOR_BUILD}" )
+    libiconv_opts+=( "ldflags=${CT_LDFLAGS_FOR_BUILD}" )
+    do_libiconv_backend "${libiconv_opts[@]}"
+
+    CT_Popd
+    CT_EndStep
+}
+
+# Build libiconv for running on host
+do_libiconv_for_host() {
+    local -a libiconv_opts
+
+    case "$CT_HOST" in
+        *darwin*|*linux*)
+            return 0
+            ;;
+    esac
+
+    CT_DoStep INFO "Installing libiconv for host"
+    CT_mkdir_pushd "${CT_BUILD_DIR}/build-libiconv-host-${CT_HOST}"
+
+    libiconv_opts+=( "host=${CT_HOST}" )
+    libiconv_opts+=( "prefix=${CT_HOST_COMPLIBS_DIR}" )
+    libiconv_opts+=( "cflags=${CT_CFLAGS_FOR_HOST}" )
+    libiconv_opts+=( "ldflags=${CT_LDFLAGS_FOR_HOST}" )
+    libiconv_opts+=( "static_build=${CT_STATIC_TOOLCHAIN}" )
+    do_libiconv_backend "${libiconv_opts[@]}"
+
+    CT_Popd
+    CT_EndStep
+}
+
+# Build libiconv
+#     Parameter     : description               : type      : default
+#     host          : machine to run on         : tuple     : (none)
+#     prefix        : prefix to install into    : dir       : (none)
+#     static_build  : build statcially          : bool      : no
+#     cflags        : host cflags to use        : string    : (empty)
+#     ldflags       : host ldflags to use       : string    : (empty)
+do_libiconv_backend() {
+    local host
+    local prefix
+    local static_build
+    local cflags
+    local ldflags
+    local arg
+    local -a extra_config
+
+    for arg in "$@"; do
+        eval "${arg// /\\ }"
+    done
+
+    CT_DoLog EXTRA "Configuring libiconv"
+
+    CT_DoExecLog ALL cp -aT "${CT_SRC_DIR}/libiconv-${CT_LIBICONV_VERSION}" "."
+
+    CT_DoExecLog CFG                                          \
+    CFLAGS="${cflags}"                                        \
+    LDFLAGS="${ldflags}"                                      \
+    "${CT_SRC_DIR}/libiconv-${CT_LIBICONV_VERSION}/configure" \
+        --build=${CT_BUILD}                                   \
+        --host="${host}"                                      \
+        --prefix="${prefix}"                                  \
+        "${extra_config[@]}"                                  \
+ 
+    CT_DoLog EXTRA "Building libiconv"
+    CT_DoExecLog ALL make CC="${CT_HOST}-gcc ${cflags}" ${JOBSFLAGS}
+
+    CT_DoLog EXTRA "Installing libiconv"
+    CT_DoExecLog ALL make install CC="${CT_HOST}-gcc ${cflags}"
+}
+
+fi
diff --git a/scripts/build/companion_libs/330-gettext.sh b/scripts/build/companion_libs/330-gettext.sh
new file mode 100644
index 0000000..e74b877
--- /dev/null
+++ b/scripts/build/companion_libs/330-gettext.sh
@@ -0,0 +1,131 @@
+# Build script for gettext
+
+do_gettext_get() { :; }
+do_gettext_extract() { :; }
+do_gettext_for_build() { :; }
+do_gettext_for_host() { :; }
+
+if [ "${CT_GETTEXT}" = "y" ]; then
+
+do_gettext_get() {
+    CT_GetFile "gettext-${CT_GETTEXT_VERSION}" \
+               http://ftp.gnu.org/pub/gnu/gettext/
+}
+
+do_gettext_extract() {
+    CT_Extract "gettext-${CT_GETTEXT_VERSION}"
+    CT_Patch "gettext" "${CT_GETTEXT_VERSION}"
+}
+
+# Build gettext for running on build
+do_gettext_for_build() {
+    local -a gettext_opts
+
+    case "$CT_BUILD" in
+        *linux*)
+            return 0
+            ;;
+    esac
+
+    CT_DoStep INFO "Installing gettext for build"
+    CT_mkdir_pushd "${CT_BUILD_DIR}/build-gettext-build-${CT_BUILD}"
+
+    gettext_opts+=( "host=${CT_BUILD}" )
+    gettext_opts+=( "prefix=${CT_BUILDTOOLS_PREFIX_DIR}" )
+    gettext_opts+=( "cflags=${CT_CFLAGS_FOR_BUILD}" )
+    gettext_opts+=( "ldflags=${CT_LDFLAGS_FOR_BUILD}" )
+    do_gettext_backend "${gettext_opts[@]}"
+
+    CT_Popd
+    CT_EndStep
+}
+
+# Build gettext for running on host
+do_gettext_for_host() {
+    local -a gettext_opts
+
+    case "$CT_HOST" in
+        *linux*)
+            return 0
+            ;;
+    esac
+
+    CT_DoStep INFO "Installing gettext for host"
+    CT_mkdir_pushd "${CT_BUILD_DIR}/build-gettext-host-${CT_HOST}"
+
+    gettext_opts+=( "host=${CT_HOST}" )
+    gettext_opts+=( "prefix=${CT_HOST_COMPLIBS_DIR}" )
+    gettext_opts+=( "cflags=${CT_CFLAGS_FOR_HOST}" )
+    gettext_opts+=( "ldflags=${CT_LDFLAGS_FOR_HOST}" )
+    gettext_opts+=( "static_build=${CT_STATIC_TOOLCHAIN}" )
+    do_gettext_backend "${gettext_opts[@]}"
+
+    CT_Popd
+    CT_EndStep
+}
+
+# Build gettext
+#     Parameter     : description               : type      : default
+#     host          : machine to run on         : tuple     : (none)
+#     prefix        : prefix to install into    : dir       : (none)
+#     static_build  : build statcially          : bool      : no
+#     cflags        : host cflags to use        : string    : (empty)
+#     ldflags       : host ldflags to use       : string    : (empty)
+do_gettext_backend() {
+    local host
+    local prefix
+    local static_build
+    local cflags
+    local ldflags
+    local arg
+    local -a extra_config
+
+    for arg in "$@"; do
+        eval "${arg// /\\ }"
+    done
+
+    CT_DoLog EXTRA "Configuring gettext"
+
+    CT_DoExecLog ALL cp -av "${CT_SRC_DIR}/gettext-${CT_GETTEXT_VERSION}/"/* .
+
+    # A bit ugly. D__USE_MINGW_ANSI_STDIO=1 has its own {v}asprintf functions
+    # but gettext configure doesn't see this flag when it checks for that. An
+    # alternative may be to use CC="${host}-gcc ${cflags}" but that didn't
+    # work.
+    # -O2 works around bug at http://savannah.gnu.org/bugs/?36443
+    # gettext needs some fixing for MinGW-w64 it would seem.
+    case "${host}" in
+        *mingw*)
+            case "${cflags}" in
+                *D__USE_MINGW_ANSI_STDIO=1*)
+                    extra_config+=( --disable-libasprintf )
+                    ;;
+            esac
+            cflags=$cflags" -O2"
+        ;;
+    esac
+
+    CT_DoExecLog CFG                                        \
+    CFLAGS="${cflags}"                                      \
+    LDFLAGS="${ldflags}"                                    \
+    "${CT_SRC_DIR}/gettext-${CT_GETTEXT_VERSION}/configure" \
+        --build=${CT_BUILD}                                 \
+        --host="${host}"                                    \
+        --prefix="${prefix}"                                \
+        --disable-java                                      \
+        --disable-native-java                               \
+        --disable-csharp                                    \
+        --enable-static                                     \
+        --enable-threads=win32                              \
+        --without-emacs                                     \
+        --disable-openmp                                    \
+        "${extra_config[@]}"
+ 
+    CT_DoLog EXTRA "Building gettext"
+    CT_DoExecLog ALL make ${JOBSFLAGS}
+
+    CT_DoLog EXTRA "Installing gettext"
+    CT_DoExecLog ALL make install
+}
+
+fi
diff --git a/scripts/build/libc/glibc-eglibc.sh-common b/scripts/build/libc/glibc-eglibc.sh-common
index 95ab33e..d388d23 100644
--- a/scripts/build/libc/glibc-eglibc.sh-common
+++ b/scripts/build/libc/glibc-eglibc.sh-common
@@ -397,10 +397,20 @@ do_libc_backend_once() {
 
         # use the 'install-headers' makefile target to install the
         # headers
-        CT_DoExecLog ALL make ${JOBSFLAGS}                          \
-                         install_root="${install_root}"             \
-                         install-bootstrap-headers=yes              \
-                         "${extra_make_args[@]}"                    \
+        case "$CT_BUILD" in
+            *mingw*)
+                # When installing headers on MinGW-w64, sunrpc needs gettext for building
+                # cross-rpcgen.
+                # used to parse the output of the C-preprocessor.
+                extra_make_args+=( BUILD_CPPFLAGS="-I${CT_BUILDTOOLS_PREFIX_DIR}/include/" )
+                extra_make_args+=( BUILD_LDFLAGS="-L${CT_BUILDTOOLS_PREFIX_DIR}/lib -lintl" )
+                ;;
+        esac
+
+        CT_DoExecLog ALL make ${JOBSFLAGS}                                        \
+                         install_root="${install_root}"                           \
+                         install-bootstrap-headers=yes                            \
+                         "${extra_make_args[@]}"                                  \
                          install-headers
 
         # For glibc, a few headers need to be manually installed
-- 
1.9.0

